// @salesforce/plugin-templates/lib/commands/visualforce/generate/page.js
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { Flags, loglevel, orgApiVersionFlagWithDeprecations, SfCommand, Ux } from '@salesforce/sf-plugins-core';
import { CreateUtil, TemplateType } from '@salesforce/templates';
import { Messages } from '@salesforce/core';
import { getCustomTemplates, runGenerator } from '../../../utils/templateCommand.js';
import { outputDirFlag } from '../../../utils/flags.js';
const visualforcePageFileSuffix = /.page$/;
const VF_TYPE = 'Page';
Messages.importMessagesDirectoryFromMetaUrl(import.meta.url);
const commonMessages = Messages.loadMessages('@salesforce/plugin-templates', 'messages');
const messages = Messages.loadMessages('@salesforce/plugin-templates', 'vf');
export default class VisualforcePage extends SfCommand {
    static summary = messages.getMessage('summary', [VF_TYPE]);
    static description = messages.getMessage('description', [VF_TYPE]);
    static examples = messages.getMessages('examples.page');
    static aliases = ['force:visualforce:page:create'];
    static deprecateAliases = true;
    static flags = {
        name: Flags.string({
            char: 'n',
            summary: messages.getMessage('flags.name.summary', [VF_TYPE]),
            description: messages.getMessage('flags.name.description'),
            required: true,
            aliases: ['pagename'],
            deprecateAliases: true,
        }),
        // there is only one valid template in the interface for VisualforcePageOptions
        template: Flags.string({
            char: 't',
            summary: commonMessages.getMessage('flags.template.summary'),
            description: commonMessages.getMessage('flags.template.description'),
            default: 'DefaultVFPage',
            hidden: true,
            options: CreateUtil.getCommandTemplatesForFiletype(visualforcePageFileSuffix, 'visualforcepage'),
        }),
        'output-dir': outputDirFlag,
        'api-version': orgApiVersionFlagWithDeprecations,
        label: Flags.string({
            char: 'l',
            summary: messages.getMessage('flags.label.summary', [VF_TYPE]),
            required: true,
        }),
        loglevel,
    };
    async run() {
        const { flags } = await this.parse(VisualforcePage);
        // translate the new flags to the old ones the generator expects
        const flagsAsOptions = {
            pagename: flags.name,
            label: flags.label,
            template: 'DefaultVFPage',
            outputdir: flags['output-dir'],
            apiversion: flags['api-version'],
        };
        return runGenerator({
            templateType: TemplateType.VisualforcePage,
            opts: flagsAsOptions,
            ux: new Ux({ jsonEnabled: this.jsonEnabled() }),
            templates: getCustomTemplates(this.configAggregator),
        });
    }
}
