// @salesforce/plugin-deploy-retrieve/lib/formatters/sourceConvertResultFormatter.js
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { resolve } from 'node:path';
import { Ux } from '@salesforce/sf-plugins-core';
import { SfError, Messages } from '@salesforce/core';
import { exitCodeAsNumber } from '../utils/output.js';
const ux = new Ux();
Messages.importMessagesDirectoryFromMetaUrl(import.meta.url);
export const convertMessages = Messages.loadMessages('@salesforce/plugin-deploy-retrieve', 'convert.source');
export class SourceConvertResultFormatter {
    result;
    constructor(result) {
        this.result = result;
    }
    // eslint-disable-next-line @typescript-eslint/require-await
    async getJson() {
        if (!this.result.packagePath) {
            throw new SfError('Convert result contains no packagePath');
        }
        return {
            location: resolve(this.result.packagePath),
        };
    }
    display() {
        if ([0, 69].includes(exitCodeAsNumber() ?? 0)) {
            ux.log(convertMessages.getMessage('success', [this.result.packagePath]));
        }
        else {
            throw new SfError(convertMessages.getMessage('convertFailed'), 'ConvertFailed');
        }
    }
}
