// @salesforce/apex-node/lib/src/reporters/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TapFormatTransformer = exports.DefaultWatermarks = exports.DefaultReportOptions = exports.CoverageReporter = exports.HumanFormatTransform = exports.HumanReporter = exports.JUnitFormatTransformer = exports.JUnitReporter = exports.TapReporter = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
var tapReporter_1 = require("./tapReporter");
Object.defineProperty(exports, "TapReporter", { enumerable: true, get: function () { return tapReporter_1.TapReporter; } });
var junitReporter_1 = require("./junitReporter");
Object.defineProperty(exports, "JUnitReporter", { enumerable: true, get: function () { return junitReporter_1.JUnitReporter; } });
var junitFormatTransformer_1 = require("./junitFormatTransformer");
Object.defineProperty(exports, "JUnitFormatTransformer", { enumerable: true, get: function () { return junitFormatTransformer_1.JUnitFormatTransformer; } });
var humanReporter_1 = require("./humanReporter");
Object.defineProperty(exports, "HumanReporter", { enumerable: true, get: function () { return humanReporter_1.HumanReporter; } });
var humanFormatTransform_1 = require("./humanFormatTransform");
Object.defineProperty(exports, "HumanFormatTransform", { enumerable: true, get: function () { return humanFormatTransform_1.HumanFormatTransform; } });
var coverageReporter_1 = require("./coverageReporter");
Object.defineProperty(exports, "CoverageReporter", { enumerable: true, get: function () { return coverageReporter_1.CoverageReporter; } });
Object.defineProperty(exports, "DefaultReportOptions", { enumerable: true, get: function () { return coverageReporter_1.DefaultReportOptions; } });
Object.defineProperty(exports, "DefaultWatermarks", { enumerable: true, get: function () { return coverageReporter_1.DefaultWatermarks; } });
var tapFormatTransform_1 = require("./tapFormatTransform");
Object.defineProperty(exports, "TapFormatTransformer", { enumerable: true, get: function () { return tapFormatTransform_1.TapFormatTransformer; } });
