// @salesforce/apex-node/lib/src/reporters/tapFormatTransform.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TapFormatTransformer = void 0;
/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const core_1 = require("@salesforce/core");
const node_stream_1 = require("node:stream");
const utils_1 = require("../utils");
const os = __importStar(require("node:os"));
class TapFormatTransformer extends node_stream_1.Readable {
    logger;
    testResult;
    epilogue;
    buffer;
    bufferSize;
    constructor(testResult, epilogue, options) {
        super(options);
        this.testResult = testResult;
        this.epilogue = epilogue;
        this.logger = core_1.Logger.childFromRoot('TapFormatTransformer');
        this.buffer = '';
        this.bufferSize = options?.bufferSize || 256; // Default buffer size is 256
    }
    pushToBuffer(chunk) {
        this.buffer += chunk;
        if (this.buffer.length >= this.bufferSize) {
            this.push(this.buffer);
            this.buffer = '';
        }
    }
    _read() {
        this.logger.trace('starting format');
        utils_1.HeapMonitor.getInstance().checkHeapSize('TapFormatTransformer._read');
        this.format();
        if (this.buffer.length > 0) {
            this.push(this.buffer);
        }
        this.push(null); // Signal the end of the stream
        this.logger.trace('finishing format');
        utils_1.HeapMonitor.getInstance().checkHeapSize('TapFormatTransformer._read');
    }
    format() {
        const testPointCount = this.testResult.tests.length;
        this.pushToBuffer(`1..${testPointCount}\n`);
        this.buildTapResults();
        this.epilogue?.forEach((c) => {
            this.pushToBuffer(`# ${c}\n`);
        });
    }
    buildTapResults() {
        this.testResult.tests.forEach((test, index) => {
            const testNumber = index + 1;
            const outcome = test.outcome === "Pass" /* ApexTestResultOutcome.Pass */ ? 'ok' : 'not ok';
            this.pushToBuffer(`${outcome} ${testNumber} ${test.fullName}\n`);
            this.buildTapDiagnostics(test).forEach((s) => {
                this.pushToBuffer(`# ${s}\n`);
            });
        });
    }
    buildTapDiagnostics(testResult) {
        const message = [];
        if (testResult.outcome !== 'Pass') {
            if (testResult.message) {
                const startsWithNewlineRegex = new RegExp(/^[/\r\n|\r|\n]\w*/gim);
                if (startsWithNewlineRegex.test(testResult.message)) {
                    testResult.message.split(/\r\n|\r|\n/g).forEach((msg) => {
                        if (msg?.length > 0) {
                            message.push(msg.trim());
                        }
                    });
                }
                else {
                    message.push(testResult.message);
                }
            }
            else {
                message.push('Unknown error');
            }
            if (testResult.stackTrace) {
                testResult.stackTrace.split(os.EOL).forEach((line) => {
                    message.push(line);
                });
            }
        }
        return message;
    }
}
exports.TapFormatTransformer = TapFormatTransformer;
__decorate([
    (0, utils_1.elapsedTime)()
], TapFormatTransformer.prototype, "format", null);
__decorate([
    (0, utils_1.elapsedTime)()
], TapFormatTransformer.prototype, "buildTapResults", null);
__decorate([
    (0, utils_1.elapsedTime)()
], TapFormatTransformer.prototype, "buildTapDiagnostics", null);
