// @salesforce/apex-node/lib/src/utils/index.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeapMonitor = exports.elapsedTime = exports.refreshAuth = exports.msToSecond = exports.formatStartTime = exports.getCurrentTime = exports.Table = exports.createFile = void 0;
var fileSystemHandler_1 = require("./fileSystemHandler");
Object.defineProperty(exports, "createFile", { enumerable: true, get: function () { return fileSystemHandler_1.createFile; } });
var table_1 = require("./table");
Object.defineProperty(exports, "Table", { enumerable: true, get: function () { return table_1.Table; } });
var dateUtil_1 = require("./dateUtil");
Object.defineProperty(exports, "getCurrentTime", { enumerable: true, get: function () { return dateUtil_1.getCurrentTime; } });
Object.defineProperty(exports, "formatStartTime", { enumerable: true, get: function () { return dateUtil_1.formatStartTime; } });
Object.defineProperty(exports, "msToSecond", { enumerable: true, get: function () { return dateUtil_1.msToSecond; } });
var authUtil_1 = require("./authUtil");
Object.defineProperty(exports, "refreshAuth", { enumerable: true, get: function () { return authUtil_1.refreshAuth; } });
var elapsedTime_1 = require("./elapsedTime");
Object.defineProperty(exports, "elapsedTime", { enumerable: true, get: function () { return elapsedTime_1.elapsedTime; } });
var heapMonitor_1 = require("./heapMonitor");
Object.defineProperty(exports, "HeapMonitor", { enumerable: true, get: function () { return heapMonitor_1.HeapMonitor; } });
