// @salesforce/templates/lib/utils/types.js
"use strict";
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.generators = exports.TemplateType = void 0;
const analyticsTemplateGenerator_1 = require("../generators/analyticsTemplateGenerator");
const apexClassGenerator_1 = require("../generators/apexClassGenerator");
const apexTriggerGenerator_1 = require("../generators/apexTriggerGenerator");
const lightningAppGenerator_1 = require("../generators/lightningAppGenerator");
const lightningComponentGenerator_1 = require("../generators/lightningComponentGenerator");
const lightningEventGenerator_1 = require("../generators/lightningEventGenerator");
const lightningInterfaceGenerator_1 = require("../generators/lightningInterfaceGenerator");
const lightningTestGenerator_1 = require("../generators/lightningTestGenerator");
const projectGenerator_1 = require("../generators/projectGenerator");
const staticResourceGenerator_1 = require("../generators/staticResourceGenerator");
const visualforceComponentGenerator_1 = require("../generators/visualforceComponentGenerator");
const visualforcePageGenerator_1 = require("../generators/visualforcePageGenerator");
/**
 * Available Template types
 * Each template type must have a corresponding generator class:
 * - generator class file should locate in generators/
 * - generator class file should default export a generator class extending SfGenerator
 * - generator class file should have a name same as the type name, except with the first letter lowercased
 */
var TemplateType;
(function (TemplateType) {
    TemplateType[TemplateType["AnalyticsTemplate"] = 0] = "AnalyticsTemplate";
    TemplateType[TemplateType["ApexClass"] = 1] = "ApexClass";
    TemplateType[TemplateType["ApexTrigger"] = 2] = "ApexTrigger";
    TemplateType[TemplateType["LightningApp"] = 3] = "LightningApp";
    TemplateType[TemplateType["LightningComponent"] = 4] = "LightningComponent";
    TemplateType[TemplateType["LightningEvent"] = 5] = "LightningEvent";
    TemplateType[TemplateType["LightningInterface"] = 6] = "LightningInterface";
    TemplateType[TemplateType["LightningTest"] = 7] = "LightningTest";
    TemplateType[TemplateType["Project"] = 8] = "Project";
    TemplateType[TemplateType["VisualforceComponent"] = 9] = "VisualforceComponent";
    TemplateType[TemplateType["VisualforcePage"] = 10] = "VisualforcePage";
    TemplateType[TemplateType["StaticResource"] = 11] = "StaticResource";
})(TemplateType || (exports.TemplateType = TemplateType = {}));
exports.generators = new Map([
    [TemplateType.AnalyticsTemplate, analyticsTemplateGenerator_1.default],
    [TemplateType.ApexClass, apexClassGenerator_1.default],
    [TemplateType.ApexTrigger, apexTriggerGenerator_1.default],
    [TemplateType.LightningApp, lightningAppGenerator_1.default],
    [TemplateType.LightningComponent, lightningComponentGenerator_1.default],
    [TemplateType.LightningEvent, lightningEventGenerator_1.default],
    [TemplateType.LightningInterface, lightningInterfaceGenerator_1.default],
    [TemplateType.LightningTest, lightningTestGenerator_1.default],
    [TemplateType.Project, projectGenerator_1.default],
    [TemplateType.StaticResource, staticResourceGenerator_1.default],
    [TemplateType.VisualforceComponent, visualforceComponentGenerator_1.default],
    [TemplateType.VisualforcePage, visualforcePageGenerator_1.default],
]);
