// @salesforce/apex-node/lib/src/streaming/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getObjectEntries = exports.getPrimitiveEntries = exports.getArrayEntries = exports.determineType = exports.TestResultStringifyStream = exports.StreamingClient = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
var streamingClient_1 = require("./streamingClient");
Object.defineProperty(exports, "StreamingClient", { enumerable: true, get: function () { return streamingClient_1.StreamingClient; } });
var testResultStringifyStream_1 = require("./testResultStringifyStream");
Object.defineProperty(exports, "TestResultStringifyStream", { enumerable: true, get: function () { return testResultStringifyStream_1.TestResultStringifyStream; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "determineType", { enumerable: true, get: function () { return utils_1.determineType; } });
Object.defineProperty(exports, "getArrayEntries", { enumerable: true, get: function () { return utils_1.getArrayEntries; } });
Object.defineProperty(exports, "getPrimitiveEntries", { enumerable: true, get: function () { return utils_1.getPrimitiveEntries; } });
Object.defineProperty(exports, "getObjectEntries", { enumerable: true, get: function () { return utils_1.getObjectEntries; } });
