// @salesforce/sf-plugins-core/lib/ux/spinner.js
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { ux } from '@oclif/core';
import { UxBase } from './base.js';
/**
 * This class is a light wrapper around ux.action that allows us to
 * automatically suppress any actions if `--json` flag is present.
 */
export class Spinner extends UxBase {
    constructor(outputEnabled) {
        super(outputEnabled);
    }
    /**
     * Get the status of the current spinner.
     */
    // eslint-disable-next-line class-methods-use-this
    get status() {
        return ux.action.status;
    }
    /**
     * Set the status of the current spinner.
     */
    // eslint-disable-next-line class-methods-use-this
    set status(status) {
        ux.action.status = status;
    }
    /**
     * Start a spinner on the console.
     */
    start(action, status, opts) {
        this.maybeNoop(() => ux.action.start(action, status, opts));
    }
    /**
     * Stop the spinner on the console.
     */
    stop(msg) {
        this.maybeNoop(() => ux.action.stop(msg));
    }
    /**
     * Pause the spinner on the console.
     */
    pause(fn, icon) {
        this.maybeNoop(() => ux.action.pause(fn, icon));
    }
}
