// @salesforce/plugin-deploy-retrieve/lib/commands/project/list/ignored.js
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { join } from 'node:path';
import fs from 'node:fs';
import { Messages, SfError } from '@salesforce/core';
import { ForceIgnore } from '@salesforce/source-deploy-retrieve';
import { SfCommand, Flags } from '@salesforce/sf-plugins-core';
Messages.importMessagesDirectoryFromMetaUrl(import.meta.url);
const messages = Messages.loadMessages('@salesforce/plugin-deploy-retrieve', 'list.ignored');
export class Ignored extends SfCommand {
    static summary = messages.getMessage('summary');
    static description = messages.getMessage('description');
    static examples = messages.getMessages('examples');
    static requiresProject = true;
    static aliases = ['force:source:ignored:list'];
    static deprecateAliases = true;
    static flags = {
        'source-dir': Flags.file({
            char: 'p',
            aliases: ['sourcepath'],
            deprecateAliases: true,
            summary: messages.getMessage('flags.source-dir.summary'),
        }),
    };
    forceIgnore;
    /**
     * Outputs all forceignored files from package directories of a project,
     * or based on a sourcepath param that points to a specific file or directory.
     */
    // eslint-disable-next-line @typescript-eslint/require-await
    async run() {
        const flags = (await this.parse(Ignored)).flags;
        try {
            this.forceIgnore = ForceIgnore.findAndCreate(this.project.getPath());
            const sourcepaths = flags['source-dir']
                ? [flags['source-dir']]
                : this.project.getUniquePackageDirectories().map((pDir) => pDir.path);
            const ignoredFiles = (await Promise.all(sourcepaths.map((sp) => this.statIgnored(sp.trim())))).flat();
            // Command output
            if (ignoredFiles.length) {
                this.log('Found the following ignored files:');
                ignoredFiles.forEach((filepath) => this.log(filepath));
            }
            else {
                this.log('No ignored files found in paths:');
                sourcepaths.forEach((sp) => this.log(sp));
            }
            return { ignoredFiles };
        }
        catch (err) {
            const error = err;
            if ('code' in error && error.code === 'ENOENT') {
                throw messages.createError('invalidSourceDir', [flags['source-dir']]);
            }
            throw SfError.wrap(error);
        }
    }
    // Stat the filepath.  Test if a file, recurse if a directory.
    async statIgnored(filepath) {
        const stats = await fs.promises.stat(filepath);
        if (stats.isDirectory()) {
            return (await Promise.all(await this.findIgnored(filepath))).flat();
        }
        else {
            return this.isIgnored(filepath) ? [filepath] : [];
        }
    }
    // Recursively search a directory for source files to test.
    async findIgnored(dir) {
        return (await fs.promises.readdir(dir)).map((filename) => this.statIgnored(join(dir, filename)));
    }
    // Test if a source file is denied, adding any ignored files to
    // the ignoredFiles array for output.
    isIgnored(filepath) {
        return this.forceIgnore.denies(filepath);
    }
}
