// @salesforce/plugin-data/lib/bulkDataRequestCache.js
/*
 * Copyright (c) 2023, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { TTLConfig, Global, Logger, Messages, Org } from '@salesforce/core';
import { Duration } from '@salesforce/kit';
Messages.importMessagesDirectoryFromMetaUrl(import.meta.url);
const messages = Messages.loadMessages('@salesforce/plugin-data', 'messages');
export class BulkDataRequestCache extends TTLConfig {
    static getDefaultOptions() {
        return {
            isGlobal: true,
            isState: true,
            filename: BulkDataRequestCache.getFileName(),
            stateFolder: Global.SF_STATE_FOLDER,
            ttl: Duration.days(3),
        };
    }
    /**
     * Creates a new bulk data request cache entry for the given bulk request id.
     *
     * @param bulkRequestId
     * @param username
     */
    async createCacheEntryForRequest(bulkRequestId, username, apiVersion) {
        if (!username) {
            throw messages.createError('usernameRequired');
        }
        this.set(bulkRequestId, {
            jobId: bulkRequestId,
            username,
            apiVersion,
        });
        await this.write();
        Logger.childFromRoot('DataRequestCache').debug(`bulk cache saved for ${bulkRequestId}`);
    }
    async resolveResumeOptionsFromCache(bulkJobId, useMostRecent, org, apiVersion) {
        if (!useMostRecent && !bulkJobId) {
            throw messages.createError('bulkRequestIdRequiredWhenNotUsingMostRecent');
        }
        const resumeOptionsOptions = {
            operation: 'query',
            query: '',
            pollingOptions: { pollTimeout: 0, pollInterval: 0 },
        };
        if (useMostRecent) {
            const key = this.getLatestKey();
            if (key) {
                // key definitely exists because it came from the cache
                const entry = this.get(key);
                return {
                    jobInfo: { id: entry.jobId },
                    options: {
                        ...resumeOptionsOptions,
                        connection: (await Org.create({ aliasOrUsername: entry.username })).getConnection(apiVersion),
                    },
                };
            }
        }
        if (bulkJobId) {
            const entry = this.get(bulkJobId);
            if (entry) {
                return {
                    jobInfo: { id: entry.jobId },
                    options: {
                        ...resumeOptionsOptions,
                        connection: (await Org.create({ aliasOrUsername: entry.username })).getConnection(apiVersion),
                    },
                };
            }
            else if (org) {
                return {
                    jobInfo: { id: bulkJobId },
                    options: {
                        ...resumeOptionsOptions,
                        connection: org.getConnection(apiVersion),
                    },
                };
            }
            else {
                throw messages.createError('cannotCreateResumeOptionsWithoutAnOrg');
            }
        }
        else if (useMostRecent) {
            throw messages.createError('cannotFindMostRecentCacheEntry');
        }
        else {
            throw messages.createError('bulkRequestIdRequiredWhenNotUsingMostRecent');
        }
    }
}
export class BulkQueryRequestCache extends BulkDataRequestCache {
    static getDefaultOptions() {
        return {
            isGlobal: true,
            isState: true,
            filename: BulkQueryRequestCache.getFileName(),
            stateFolder: Global.SF_STATE_FOLDER,
            ttl: Duration.days(3),
        };
    }
    static getFileName() {
        return 'bulk-data-query-cache.json';
    }
    static async unset(key) {
        const cache = await BulkQueryRequestCache.create();
        cache.unset(key);
        await cache.write();
    }
}
export class BulkDeleteRequestCache extends BulkDataRequestCache {
    static getDefaultOptions() {
        return {
            isGlobal: true,
            isState: true,
            filename: BulkDeleteRequestCache.getFileName(),
            stateFolder: Global.SF_STATE_FOLDER,
            ttl: Duration.days(3),
        };
    }
    static getFileName() {
        return 'bulk-data-delete-cache.json';
    }
    static async unset(key) {
        const cache = await BulkDeleteRequestCache.create();
        cache.unset(key);
        await cache.write();
    }
}
export class BulkUpsertRequestCache extends BulkDataRequestCache {
    static getDefaultOptions() {
        return {
            isGlobal: true,
            isState: true,
            filename: BulkUpsertRequestCache.getFileName(),
            stateFolder: Global.SF_STATE_FOLDER,
            ttl: Duration.days(3),
        };
    }
    static getFileName() {
        return 'bulk-data-upsert-cache.json';
    }
    static async unset(key) {
        const cache = await BulkUpsertRequestCache.create();
        cache.unset(key);
        await cache.write();
    }
}
