// @salesforce/plugin-data/lib/types.js
export var FieldType;
(function (FieldType) {
    FieldType[FieldType["field"] = 0] = "field";
    FieldType[FieldType["subqueryField"] = 1] = "subqueryField";
    FieldType[FieldType["functionField"] = 2] = "functionField";
})(FieldType || (FieldType = {}));
/** element: a field value from the sobject.  Empty arrays return true */
export const hasNestedRecords = (element) => Array.isArray(element?.records);
/** element: a field value from the sobject.  Empty arrays return false */
export const hasNonEmptyNestedRecords = (element) => hasNestedRecords(element) && element.records.length > 0;
/** convenience method for filtering Object.entries array */
export const hasNestedRecordsFilter = (entry) => typeof entry[0] === 'string' && hasNonEmptyNestedRecords(entry[1]);
export const isAttributesElement = (element) => !!element.referenceId && !!element.type;
/** convenience method for filtering Object.entries array */
export const isAttributesEntry = (entry) => entry[0] === 'attributes' && isAttributesElement(entry[1]);
